In [1]:
import numpy as np
import pandas as pd

In [2]:
books=pd.read_csv('BX-Books.csv',sep=';',error_bad_lines=False,encoding='latin-1')#latin-1 is actually ISO-8859-1

b'Skipping line 6452: expected 8 fields, saw 9\nSkipping line 43667: expected 8 fields, saw 10\nSkipping line 51751: expected 8 fields, saw 9\n'
b'Skipping line 92038: expected 8 fields, saw 9\nSkipping line 104319: expected 8 fields, saw 9\nSkipping line 121768: expected 8 fields, saw 9\n'
b'Skipping line 144058: expected 8 fields, saw 9\nSkipping line 150789: expected 8 fields, saw 9\nSkipping line 157128: expected 8 fields, saw 9\nSkipping line 180189: expected 8 fields, saw 9\nSkipping line 185738: expected 8 fields, saw 9\n'
b'Skipping line 209388: expected 8 fields, saw 9\nSkipping line 220626: expected 8 fields, saw 9\nSkipping line 227933: expected 8 fields, saw 11\nSkipping line 228957: expected 8 fields, saw 10\nSkipping line 245933: expected 8 fields, saw 9\nSkipping line 251296: expected 8 fields, saw 9\nSkipping line 259941: expected 8 fields, saw 9\nSkipping line 261529: expected 8 fields, saw 9\n'
C:\Users\bhasw\anaconda3\lib\site-packages\IPython\core\interactiveshell.p

In [3]:
books.head()

ISBN                                         Book-Title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

            Book-Author Year-Of-Publication                   Publisher  \
0    Mark P. O. Morford                2002     Oxford University Press   
1  Richard Bruce Wright                2001       HarperFlamingo Canada   
2          Carlo D'Este                1991             HarperPerennial   
3      Gina Bari Kolata                1999        Farrar Straus Giroux   
4       E. J. W. Barber                1999  W. W. Norton &amp; Company   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

In [4]:
books=books[['ISBN','Book-Title','Book-Author','Year-Of-Publication','Publisher']]

In [5]:
books.head()

ISBN                                         Book-Title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

            Book-Author Year-Of-Publication                   Publisher  
0    Mark P. O. Morford                2002     Oxford University Press  
1  Richard Bruce Wright                2001       HarperFlamingo Canada  
2          Carlo D'Este                1991             HarperPerennial  
3      Gina Bari Kolata                1999        Farrar Straus Giroux  
4       E. J. W. Barber                1999  W. W. Norton &amp; Company

In [6]:
books.rename(columns={'Book-Title':'title','Book-Author':'author','Year-Of-Publication':'year','Publisher':'publisher'},inplace=True)

In [7]:
books.head(2)

ISBN                title                author  year  \
0  0195153448  Classical Mythology    Mark P. O. Morford  2002   
1  0002005018         Clara Callan  Richard Bruce Wright  2001   

                 publisher  
0  Oxford University Press  
1    HarperFlamingo Canada

In [8]:
users=pd.read_csv('BX-Users.csv',sep=';',error_bad_lines=False,encoding='latin-1')

In [9]:
users.head(2)

User-ID                   Location   Age
0        1         nyc, new york, usa   NaN
1        2  stockton, california, usa  18.0

In [10]:
users.rename(columns={'User-ID':'user_id','Location':'location','Age':'age'},inplace=True)

In [11]:
users.head(2)

user_id                   location   age
0        1         nyc, new york, usa   NaN
1        2  stockton, california, usa  18.0

In [12]:
ratings=pd.read_csv('BX-Book-Ratings.csv',sep=';',error_bad_lines=False,encoding='latin-1')

In [13]:
ratings.head(2)

User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5

In [14]:
ratings.rename(columns={'User-ID':'user_id','Book-Rating':'rating'},inplace=True)

In [15]:
ratings.head(2)

user_id        ISBN  rating
0   276725  034545104X       0
1   276726  0155061224       5

In [16]:
books.shape

(271360, 5)

In [17]:
users.shape

(278858, 3)

In [18]:
ratings.shape

(1149780, 3)

### Create a table which has user_id as columns and books as rows or indexes. Only those user_id are taken who has given rating to more than 200 books and those books are taken which has been rated by more than 50 users.Then only it will recommend good books.

In [19]:
ratings['user_id'].value_counts()

11676     13602
198711     7550
153662     6109
98391      5891
35859      5850
          ...  
158698        1
17920         1
277135        1
275086        1
187812        1
Name: user_id, Length: 105283, dtype: int64

In [20]:
ratings['user_id'].value_counts().shape

(105283,)

In [21]:
x=ratings['user_id'].value_counts()>=200

In [22]:
x[x] #only True values are present

11676     True
198711    True
153662    True
98391     True
35859     True
          ... 
225595    True
252827    True
99955     True
36554     True
83671     True
Name: user_id, Length: 905, dtype: bool

In [23]:
x[x].shape

(905,)

In [24]:
y=x[x].index

In [25]:
ratings=ratings[ratings['user_id'].isin(y)] #only those rows are selected whose user_id are present in y

In [26]:
ratings.shape

(527556, 3)

In [27]:
ratings.head()

user_id        ISBN  rating
1456   277427  002542730X      10
1457   277427  0026217457       0
1458   277427  003008685X       8
1459   277427  0030615321       0
1460   277427  0060002050       0

In [28]:
ratings_with_books=ratings.merge(books,on='ISBN')

In [29]:
ratings_with_books.shape

(488742, 7)

In [30]:
number_rating=ratings_with_books.groupby('title')['rating'].count().reset_index()

In [31]:
number_rating.rename(columns={'rating':'number_of_ratings'},inplace=True)

In [32]:
number_rating

title  number_of_ratings
0        A Light in the Storm: The Civil War Diary of ...                  2
1                                   Always Have Popsicles                  1
2                    Apple Magic (The Collector's series)                  1
3        Beyond IBM: Leadership Marketing and Finance ...                  1
4        Clifford Visita El Hospital (Clifford El Gran...                  1
...                                                   ...                ...
160571  Ã?Â?ber die Pflicht zum Ungehorsam gegen den S...                  3
160572                                      Ã?Â?lpiraten.                  1
160573                     Ã?Â?rger mit Produkt X. Roman.                  1
160574                              Ã?Â?stlich der Berge.                  1
160575                                  Ã?Â?thique en toc                  1

[160576 rows x 2 columns]

In [33]:
final_rating=ratings_with_books.merge(number_rating,on='title')

In [34]:
final_rating=final_rating[final_rating['number_of_ratings']>=50]

In [35]:
final_rating.shape

(62149, 8)

In [36]:
final_rating.drop_duplicates(['user_id','title'],inplace=True)

In [37]:
final_rating.shape

(60141, 8)

### So there are 60141 unique ratings which has been done by those users who has given ratings to more than 200 books and those books are present which has got ratings of more than 50 users

In [38]:
book_pivot=final_rating.pivot_table(columns='user_id',index='title',values='rating')

In [39]:
book_pivot

user_id                                             254     2276    2766    \
title                                                                        
1984                                                   9.0     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN    10.0     NaN   
4 Blondes                                              NaN     NaN     NaN   
84 Charing Cross Road                                  NaN     NaN     NaN   
...                                                    ...     ...     ...   
Year of Wonders                                        NaN     NaN     NaN   
You Belong To Me                                       NaN     NaN     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     NaN     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

user_id                                             2977    3363    3757    \
title                                                                        
1984                                                   NaN     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN     NaN     NaN   
4 Blondes                                              NaN     NaN     NaN   
84 Charing Cross Road                                  NaN     NaN     NaN   
...                                                    ...     ...     ...   
Year of Wonders                                        7.0     NaN     NaN   
You Belong To Me                                       NaN     NaN     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     0.0     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

user_id                                             4017    4385    6242    \
title                                                                        
1984                                                   NaN     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN     NaN     NaN   
4 Blondes                                              NaN     NaN     NaN   
84 Charing Cross Road                                  NaN     NaN     NaN   
...                                                    ...     ...     ...   
Year of Wonders                                        NaN     NaN     7.0   
You Belong To Me                                       NaN     NaN     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     NaN     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

user_id                                             6251    ...  274004  \
title                                                       ...           
1984                                                   NaN  ...     NaN   
1st to Die: A Novel                                    NaN  ...     NaN   
2nd Chance                                             NaN  ...     NaN   
4 Blondes                                              0.0  ...     NaN   
84 Charing Cross Road                                  NaN  ...     NaN   
...                                                    ...  ...     ...   
Year of Wonders                                        NaN  ...     NaN   
You Belong To Me                                       NaN  ...     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     0.0  ...     NaN   
Zoya                                                   NaN  ...     NaN   
\O\" Is for Outlaw"                                    

In [40]:
book_pivot.shape

(746, 894)

In [41]:
book_pivot.fillna(0,inplace=True)

In [42]:
book_pivot

user_id                                             254     2276    2766    \
title                                                                        
1984                                                   9.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0    10.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
84 Charing Cross Road                                  0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

user_id                                             2977    3363    3757    \
title                                                                        
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
84 Charing Cross Road                                  0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        7.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

user_id                                             4017    4385    6242    \
title                                                                        
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
84 Charing Cross Road                                  0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     7.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

user_id                                             6251    ...  274004  \
title                                                       ...           
1984                                                   0.0  ...     0.0   
1st to Die: A Novel                                    0.0  ...     0.0   
2nd Chance                                             0.0  ...     0.0   
4 Blondes                                              0.0  ...     0.0   
84 Charing Cross Road                                  0.0  ...     0.0   
...                                                    ...  ...     ...   
Year of Wonders                                        0.0  ...     0.0   
You Belong To Me                                       0.0  ...     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0  ...     0.0   
Zoya                                                   0.0  ...     0.0   
\O\" Is for Outlaw"                                    

### Now we will use nearest neighbor algo which is a clustering algo and it is just like KNN which is a classfication algo.So in nearest neighbor algo we will find out the distances and put the most closest point in one cluster

### For this calculation of distances , we don't need 0's so to remove those we will convert the pivot table into sparse matrix

In [43]:
from scipy.sparse import csr_matrix
book_sparse=csr_matrix(book_pivot)

### So book_sparse is a csr_matrix where non zero values are considered

In [44]:
from sklearn.neighbors import NearestNeighbors
model=NearestNeighbors(algorithm='brute') #brute means to calculate distances of every point from every other points and then consider only the nearest points and put in a cluster

In [45]:
model.fit(book_sparse)

NearestNeighbors(algorithm='brute')

In [46]:
distances,suggestions=model.kneighbors(book_pivot.iloc[237,:].values.reshape(1,-1),n_neighbors=6)

In [47]:
distances

array([[ 0.        , 68.78953409, 70.        , 72.64296249, 76.83098333,
        77.28518616]])

In [48]:
suggestions

array([[237, 240, 238, 241, 184, 539]], dtype=int64)

In [49]:
book_pivot.index[237]

'Harry Potter and the Chamber of Secrets (Book 2)'

In [50]:
for i in range(len(suggestions)):
    print(book_pivot.index[suggestions[i]])

Index(['Harry Potter and the Chamber of Secrets (Book 2)',
       'Harry Potter and the Prisoner of Azkaban (Book 3)',
       'Harry Potter and the Goblet of Fire (Book 4)',
       'Harry Potter and the Sorcerer's Stone (Book 1)', 'Exclusive',
       'The Cradle Will Fall'],
      dtype='object', name='title')


In [51]:
np.where(book_pivot.index=='Animal Farm')[0][0]

54

In [52]:
def recommend_book(book_name):
    book_id=np.where(book_pivot.index==book_name)[0][0]
    distances,suggestions=model.kneighbors(book_pivot.iloc[book_id,:].values.reshape(1,-1),n_neighbors=6)

    x=[]
    
    for i in range(len(suggestions)):
        for j in book_pivot.index[suggestions[i]]:
            x.append(j)
            
    
    y=[]
  
    for i in range(1,len(x)):
        print(x[i])
        
    

In [53]:
recommend_book('Animal Farm')

Exclusive
Jacob Have I Loved
Second Nature
Pleading Guilty
No Safe Place


In [54]:
final_rating.to_csv('Final_Rating.csv')